app/src/main/java/de/westnordost/streetcomplete/quests/building_type/BuildingType.kt

```diff
@@ -0,0 +1,121 @@
+    STATIC_CARAVAN("building", "static_caravan"),
+    HUT           ("building", "hut"),
+
+    INDUSTRIAL    ("building", "industrial"),
+    RETAIL        ("building", "retail"),
+    OFFICE        ("building", "office"),
+    WAREHOUSE     ("building", "warehouse"),
+    KIOSK         ("building", "kiosk"),
+    STORAGE_TANK  ("man_made", "storage_tank"),
+
+    KINDERGARTEN  ("building", "kindergarten"),
+    SCHOOL        ("building", "school"),
+    COLLEGE       ("building", "college"),
+    SPORTS_CENTRE ("building", "sports_centre"),
+    HOSPITAL      ("building", "hospital"),
+    STADIUM       ("building", "stadium"),
+    TRAIN_STATION ("building", "train_station"),
+    TRANSPORTATION("building", "transportation"),
+    UNIVERSITY    ("building", "university"),
+    GOVERNMENT    ("building", "government"),
+
+    CHURCH        ("building", "church"),
+    CHAPEL        ("building", "chapel"),
+    CATHEDRAL     ("building", "cathedral"),
+    MOSQUE        ("building", "mosque"),
+    TEMPLE        ("building", "temple"),
+    PAGODA        ("building", "pagoda"),
+    SYNAGOGUE     ("building", "synagogue"),
+    SHRINE        ("building", "shrine"),
+
+    CARPORT       ("building", "carport"),
+    GARAGE        ("building", "garage"),
+    GARAGES       ("building", "garages"),
+    PARKING       ("building", "parking"),
+
+    FARM          ("building", "farm"),
+    FARM_AUXILIARY("building", "farm_auxiliary"),
+    SILO          ("man_made", "silo"),
+    GREENHOUSE    ("building", "greenhouse"),
+
+    SHED          ("building", "shed"),
+    ROOF          ("building", "roof"),
+    TOILETS       ("building", "toilets"),
+    SERVICE       ("building", "service"),
+    HANGAR        ("building", "hangar"),
+    BUNKER        ("building", "bunker"),
+    HISTORIC      ("historic", "yes"),
+    ABANDONED     ("abandoned", "yes"),
+    RUINS         ("ruins", "yes"),
+
+    RESIDENTIAL   ("building", "residential"),
+    COMMERCIAL    ("building", "commercial"),
+    CIVIC         ("building", "civic"),
+    RELIGIOUS     ("building", "religious"),
+
+    CONSTRUCTION ("building", "construction");
+
+    companion object {
+        fun getByTag(key: String, value: String): BuildingType? {
+            return values().find { it.osmKey == key && it.osmValue == value }
+        }
+    }
+}
+
+val topBuildingTypes = listOf(
+    BuildingType.DETACHED, BuildingType.APARTMENTS, BuildingType.HOUSE, BuildingType.GARAGE,
+    BuildingType.SHED, BuildingType.HUT
+)
+
+enum class BuildingTypeCategory(val type: BuildingType?, val subTypes: List<BuildingType>) {
+    RESIDENTIAL(BuildingType.RESIDENTIAL, listOf(
+        BuildingType.DETACHED, BuildingType.APARTMENTS, BuildingType.SEMI_DETACHED,
+        BuildingType.TERRACE, BuildingType.HOUSE, BuildingType.FARM, BuildingType.HUT,
+        BuildingType.BUNGALOW, BuildingType.HOUSEBOAT, BuildingType.STATIC_CARAVAN,
+        BuildingType.DORMITORY
+    )),
```

Comment:
Interesting solution, nice work!

Maybe it would improve readability if you could cut out the `BuildingType.*` in this file.

It should work if you 
```kotlin
import de.westnordost.streetcomplete.quests.building_type.BuildingType.*
import de.westnordost.streetcomplete.quests.building_type.BuildingTypeCategory.*
```

Reply:
I don't think that will work — `BuildingType` is already declared in the same file.

However, I think a scope function should work:

```kotlin
    RESIDENTIAL(BuildingType.RESIDENTIAL, with(BuildingType) { listOf(
        DETACHED, APARTMENTS, SEMI_DETACHED, TERRACE, HOUSE, FARM, HUT,
        BUNGALOW, HOUSEBOAT, STATIC_CARAVAN, DORMITORY
    )}),
```

Didn't test this; don't take my word that it will work.

Reply:
Personally I would prefer this even if BuildingType were declared in a different package — as someone who seldom uses an IDE, wildcard imports make it a real pain to figure out where something comes from.

Reply:
@smichel17 Unfortunately, your suggestion doesn't work:

```
…/quests/building_type/BuildingType.kt: (84, 9): Unresolved reference: DETACHED
```

Reply:
Ah, that's sad. …but the import trick did work! I'm surprised — didn't realize you could import a package from within itself.

Reply:
For posterity, this would also work (tested this time), although I prefer the wildcard import given the definitions are contained in the same file
```kotlin
listOf("STATIC_CARAVAN", "DORMITORY").map(BuildingType::valueOf)
```

Reply:
> listOf("STATIC_CARAVAN", "DORMITORY").map(BuildingType::valueOf)

Goodby compile time safety, hello RuntimeException

Reply:
I call it a [compiler bug](https://www.staging-typescript.org/docs/handbook/literal-types.html#string-literal-types)